### ISSUE TITLE
Address matching, glob / wildcards

### ISSUE THREAD

**Issue by avlapp** [scgolang/osc#15]
Is there support for OSC address pattern wildcards like '*', '?', '{,}' and '[]'? 

(Will do some testing later.)

**Comment by avlapp** [comment]
I don't think so, it's the main thing I'm missing in this library at the moment. (help wanted). ;)

**Comment by vlappa** [comment]
It looks like this is a misunderstanding of the OSC spec:

var invalidAddressRunes = []rune{'*', '?', ',', '[', ']', '{', '}', '#', ' '}

"// ValidateAddress returns an error if addr contains
// characters that are disallowed by the OSC spec."

https://github.com/scgolang/osc/blob/ba4a707618ec302feb8a03997d1cc6778c1c7f5f/conn.go#L34

I think the OSC spec doesn't talk about OSC addresses, but about those "printable ASCII characters not allowed in names of OSC Methods or OSC Containers"

https://opensoundcontrol.stanford.edu/spec-1_0.html

**Comment by vlappa** [comment]
It looks like this is a misunderstanding of the OSC spec:

var invalidAddressRunes = []rune{'*', '?', ',', '[', ']', '{', '}', '#', ' '}

"// ValidateAddress returns an error if addr contains
// characters that are disallowed by the OSC spec."

https://github.com/scgolang/osc/blob/ba4a707618ec302feb8a03997d1cc6778c1c7f5f/conn.go#L34

I think the OSC spec doesn't talk about OSC addresses, but about those "printable ASCII characters not allowed in names of OSC Methods or OSC Containers"

https://opensoundcontrol.stanford.edu/spec-1_0.html

**Comment by KopiasCsaba** [comment]
I've found the same issue, and I've found the soultion too, I think it's a simple accident:

Here:
https://github.com/scgolang/osc/blob/53342a90504aa5723f1044cd9ffc25e6b962cee5/message.go#L92

``xp, err := GetRegex(msg.Address)`

should be just

`xp, err := GetRegex(address)`


When 'ill ihave time I might open a PR on that.

**Comment by vlappa** [comment]
Finally someone else who seems to use this library. :) If that works, great find. Yes please a PR would be welcome.

My biggest practical issue with this library is currently this one:
https://github.com/scgolang/osc/issues/17

I've to use a external tool which uses the hypebeast/go-osc library to send a empty string (to a client that uses the C/C++ liblo library).

Would be nice if these issues can be sorted out, in general this library is really useful for me.

**Comment by KopiasCsaba** [comment]
I've tried to use it, but it sends extra garbage along in the commands, and my behringer x32 doesn't like it:(

**Comment by vlappa** [comment]
https://github.com/scgolang/osc/blob/ba4a707618ec302feb8a03997d1cc6778c1c7f5f/conn.go#L32

It doesn't accept globs at all

It indeed attempts to support Regex, so my second msg in this thread is a wrong assumption.

Looking at this func:
https://github.com/scgolang/osc/blob/53342a90504aa5723f1044cd9ffc25e6b962cee5/message.go#L147

It seems to expect that both addresses have the same amount of parts, maybe there is goes wrong?

If the address in the handler is /foo/bar/one

/foo/* wouldn't work, cause it doesn't have the same number of parts, while /foo/*/* should work I think.

**Comment by KopiasCsaba** [comment]
> It seems to expect that both addresses have the same amount of parts, maybe there it goes wrong?

Thats true, so the message address should have the same number of parts (splitted by /), and then the pattern matching occurs.

So saying "*" will only match /foo or /bar, not /foo/bar.

So then the pattern matching had a little mixup, for which I just opened the PR to fix.

**Comment by vlappa** [comment]
I doubt this is what 'we' want. I saw @briansorahan was also doing stuff around NSM and for broadcasting it seems it has to accept all paths. So it would be good if this library would support this. 

From nsm.h which uses the c/c++ liblo library:

lo_server_add_method( _NSM()->_server, "/error", "sis", _nsm_osc_error, _NSM() );
    lo_server_add_method( _NSM()->_server, "/reply", "ssss", _nsm_osc_announce_reply, _NSM() );
    lo_server_add_method( _NSM()->_server, "/nsm/client/open", "sss", _nsm_osc_open, _NSM() );
    lo_server_add_method( _NSM()->_server, "/nsm/client/save", "", _nsm_osc_save, _NSM() );
    lo_server_add_method( _NSM()->_server, "/nsm/client/session_is_loaded", "", _nsm_osc_session_is_loaded, _NSM() );
    lo_server_add_method( _NSM()->_server, NULL, NULL, _nsm_osc_broadcast, _NSM() );


_path | The OSC path to register the method to. If NULL is passed the method will match all paths._
https://liblo.sourceforge.net/docs/group__liblolowlevel.html#ga56c5f1d587c215c27e92e6436e12b1a3

**Comment by KopiasCsaba** [comment]
Yes, you are right, I have a PR on adding a regexp matcher already, see: https://github.com/scgolang/osc/pull/20!
So the PatternMatcher is the standard compilant one (after my other PR is merged), but we indeed need something else.
 
But, you can make your own meanwhile just implement the Dispatcher interface!

**Comment by KopiasCsaba** [comment]
(Or you can copy from my code, or customize the one from here: https://github.com/scgolang/osc/blob/master/customdispatcher_example_test.go)

**Comment by vlappa** [comment]
Thanks, I'll look at it more closely when I've more time.
